### DISCUSSION TITLE
Local and global components

### DISCUSSION THREAD

**Discussion by TreyWW** [django-components/django-components#333] [Q&A]
I've recently been using django-components and they're really cool. But i want to have a seperation between `global components` and `local components`. Where local components are almost "included" or only used for a particular page. 

For example
Global components may be: Modal, message_list
Local components may be: my_profile_icon, tabs

At the moment, i made a custom extension of django-components which automatically creates any components with a prefix that are inside of /components/. These are "global" components we could call them.
And then any in /frontend/templates/ are "local" but to access these you have to use the component name like: pages:page_name_group:page_name:component

I want to almost have the same name as a global one, but thats just locally imported. So like
**"local:component"** which can only be used in "registered" views. Any ideas for adding this kind of functionality?

I'm happy to provide more details or explanation if that didn't make sense. 
Thanks :) 

Edit: as for the local components, they'll be in the same folder as normal pages. So like: `frontend/pages/page group/page name/+component.html`
And then from there that needs to somehow create a locally importable class or something that I can put in my view that renders to the html page?

**Comment by EmilStenstrom** [comment]
Hi @TreyWW! I'm not sure I follow. First it sounds like you have already solved you problem with a custom extension. They it sounds like the problem with your solution is that the names are too long, and you want to shorten dem. Then there's the part with a "registered view", what's that? Could you try show two examples of how you would like to import a local and global components, and what the differences are?

**Reply by TreyWW** [reply]
Yeah sure thing, i did kind of think my explanation wouldn't make sense lol, even I didn't follow most of it.

Okay so, at the moment for my "local" components the names get created automatically as like lets say:
```html
{% component "frontend:pages:mypage:mylocalcomponent" question=flashcard.0 answer=flashcard.1 %}
```

But with local components, you'd go into your view, and maybe add a decorator, or a context

```py
@component.register("frontend/pages/mypage/component1", "mylocalcomponent")
def my_view(request: HttpRequest):
    return render(request, "mypage.html", {})
```
then with your html template, now you can just do

```html
{% component local:mylocalcomponent" question=flashcard.0 answer=flashcard.1 %}
```

This in my opinion would be much easier to write, understand and read. And it's local, so not every page will have this. So it's more tailored to that view.

So my "script" lets say just scans through my codebase for any html files with the prefix, and adds a component class for that. So like

`/frontend/pages/page_group/page/+component` turns into 
```html
{% component "page_group:page:mylocalcomponent" %}
```

So indeed you could argue, it's short enough as it is... which im fine with at the moment, but for concept sake, and overall shortening i do believe local components would be cool.

I just wanted a little point in the direction of how I may be able to do this, if you don't fancy adding this (makes complete sense if you don't, its a good feature but may be big?).

If this isn't added, my next plan would likely to be to  edit the script so any files that are in /frontend/ are registered with a different name system, so maybe like local:page_name:component. just wont be local, itll just be shortened versions of my current global ones


Thanks for responding 😄
